### PR TITLE
Extract SignalHandler class for re-use

### PR DESCRIPTION
**Pull request by oschaaf**
In preparation of sharing functionality for signal handling,
extract what we have right now into a `SignalHandler` class.

Preparation for #280

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by oschaaf** [comment]
@mum4k requesting early feedback on this; this extracts our signal handling code from `nighthawk_service` into a separate class for re-use later. In terms of design and fool proofing this we can probably enhance this, but a perfect design is not the initial goal: 
- I think this would be an improvement compared to what's on master today as we contain this functionality.
- `service_main.cc` gets much cleaner.

Would it make sense to shoot for landing this as-is?

File: source/common/signal_handler.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+#include <functional>
+#include <memory>
+#include <thread>
+#include <vector>
+
+#include "external/envoy/source/common/common/logger.h"
+
+namespace Nighthawk {
+
+using SignalCallback = std::function<void()>;
```
Comment (mum4k) [inline source/common/signal_handler.h:12]: Since this is part of the header - can we document it?

File: source/common/signal_handler.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+class SignalHandler final : public Envoy::Logger::Loggable<Envoy::Logger::Id::main> {
```
Comment (mum4k) [inline source/common/signal_handler.h:14]: Let's also add documentation for the class including an example of usage.

File: source/common/signal_handler.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+public:
+  SignalHandler(const SignalCallback& signal_callback);
```
Comment (mum4k) [inline source/common/signal_handler.h:16]: Can we add documentation for the constructor?

File: source/common/signal_handler.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  SignalHandler(SignalHandler const&) = delete;
```
Comment (mum4k) [inline source/common/signal_handler.h:17]: Can we add a single comment above these two to indicate the intention, something like:

// Not copyable or movable.

**Comment by oschaaf** [comment]
@mum4k thanks, comments addressed in e665347

**Pull request merged**